### PR TITLE
restore: fix #140 lightning breaks on tidb-master

### LINKED ISSUE #140: lightning breaks on tidb-master
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do? If possible, provide a recipe for reproducing the error.

I have a script to download `master` daily, and then import the ontime database into TiDB.  It just started breaking because of https://github.com/pingcap/parser/pull/231 which was backported to 2.1 as https://github.com/pingcap/parser/pull/232

```
#!/bin/sh
mysql -e 'drop database if exists ontime'
mysql -e 'create database ontime'
mysql ontime -e "CREATE TABLE ontime (
  Year year(4) DEFAULT NULL,
  Quarter tinyint(4) DEFAULT NULL,
  Month tinyint(4) DEFAULT NULL,
  DayofMonth tinyint(4) DEFAULT NULL,
  DayOfWeek tinyint(4) DEFAULT NULL,
  FlightDate date DEFAULT NULL,
  UniqueCarrier char(7) DEFAULT NULL,
  AirlineID int(11) DEFAULT NULL,
  Carrier char(2) DEFAULT NULL,
  TailNum varchar(50) DEFAULT NULL,
  FlightNum varchar(10) DEFAULT NULL,
  OriginAirportID int(11) DEFAULT NULL,
  OriginAirportSeqID int(11) DEFAULT NULL,
  OriginCityMarketID int(11) DEFAULT NULL,
  Origin char(5) DEFAULT NULL,
  OriginCityName varchar(100) DEFAULT NULL,
  OriginState char(2) DEFAULT NULL,
  OriginStateFips varchar(10) DEFAULT NULL,
  OriginStateName varchar(100) DEFAULT NULL,
  OriginWac int(11) DEFAULT NULL,
  DestAirportID int(11) DEFAULT NULL,
  DestAirportSeqID int(11) DEFAULT NULL,
  DestCityMarketID int(11) DEFAULT NULL,
  Dest char(5) DEFAULT NULL,
  DestCityName varchar(100) DEFAULT NULL,
  DestState char(2) DEFAULT NULL,
  DestStateFips varchar(10) DEFAULT NULL,
  DestStateName varchar(100) DEFAULT NULL,
  DestWac int(11) DEFAULT NULL,
  CRSDepTime int(11) DEFAULT NULL,
  DepTime int(11) DEFAULT NULL,
  DepDelay int(11) DEFAULT NULL,
  DepDelayMinutes int(11) DEFAULT NULL,
  DepDel15 int(11) DEFAULT NULL,
  DepartureDelayGroups int(11) DEFAULT NULL,
  DepTimeBlk varchar(20) DEFAULT NULL,
  TaxiOut int(11) DEFAULT NULL,
  WheelsOff int(11) DEFAULT NULL,
  WheelsOn int(11) DEFAULT NULL,
  TaxiIn int(11) DEFAULT NULL,
  CRSArrTime int(11) DEFAULT NULL,
  ArrTime int(11) DEFAULT NULL,
  ArrDelay int(11) DEFAULT NULL,
  ArrDelayMinutes int(11) DEFAULT NULL,
  ArrDel15 int(11) DEFAULT NULL,
  ArrivalDelayGroups int(11) DEFAULT NULL,
  ArrTimeBlk varchar(20) DEFAULT NULL,
  Cancelled tinyint(4) DEFAULT NULL,
  CancellationCode char(1) DEFAULT NULL,
  Diverted tinyint(4) DEFAULT NULL,
  CRSElapsedTime int(11) DEFAULT NULL,
  ActualElapsedTime int(11) DEFAULT NULL,
  AirTime int(11) DEFAULT NULL,
  Flights int(11) DEFAULT NULL,
  Distance int(11) DEFAULT NULL,
  DistanceGroup tinyint(4) DEFAULT NULL,
  CarrierDelay int(11) DEFAULT NULL,
  WeatherDelay int(11) DEFAULT NULL,
  NASDelay int(11) DEFAULT NULL,
  SecurityDelay int(11) DEFAULT NULL,
  LateAircraftDelay int(11) DEFAULT NULL,
  FirstDepTime varchar(10) DEFAULT NULL,
  TotalAddGTime varchar(10) DEFAULT NULL,
  LongestAddGTime varchar(10) DEFAULT NULL,
  DivAirportLandings varchar(10) DEFAULT NULL,
  DivReachedDest varchar(10) DEFAULT NULL,
  DivActualElapsedTime varchar(10) DEFAULT NULL,
  DivArrDelay varchar(10) DEFAULT NULL,
  DivDistance varchar(10) DEFAULT NULL,
  Div1Airport varchar(10) DEFAULT NULL,
  Div1AirportID int(11) DEFAULT NULL,
  Div1AirportSeqID int(11) DEFAULT NULL,
  Div1WheelsOn varchar(10) DEFAULT NULL,
  Div1TotalGTime varchar(10) DEFAULT NULL,
  Div1LongestGTime varchar(10) DEFAULT NULL,
  Div1WheelsOff varchar(10) DEFAULT NULL,
  Div1TailNum varchar(10) DEFAULT NULL,
  Div2Airport varchar(10) DEFAULT NULL,
  Div2AirportID int(11) DEFAULT NULL,
  Div2AirportSeqID int(11) DEFAULT NULL,
  Div2WheelsOn varchar(10) DEFAULT NULL,
  Div2TotalGTime varchar(10) DEFAULT NULL,
  Div2LongestGTime varchar(10) DEFAULT NULL,
  Div2WheelsOff varchar(10) DEFAULT NULL,
  Div2TailNum varchar(10) DEFAULT NULL,
  Div3Airport varchar(10) DEFAULT NULL,
  Div3AirportID int(11) DEFAULT NULL,
  Div3AirportSeqID int(11) DEFAULT NULL,
  Div3WheelsOn varchar(10) DEFAULT NULL,
  Div3TotalGTime varchar(10) DEFAULT NULL,
  Div3LongestGTime varchar(10) DEFAULT NULL,
  Div3WheelsOff varchar(10) DEFAULT NULL,
  Div3TailNum varchar(10) DEFAULT NULL,
  Div4Airport varchar(10) DEFAULT NULL,
  Div4AirportID int(11) DEFAULT NULL,
  Div4AirportSeqID int(11) DEFAULT NULL,
  Div4WheelsOn varchar(10) DEFAULT NULL,
  Div4TotalGTime varchar(10) DEFAULT NULL,
  Div4LongestGTime varchar(10) DEFAULT NULL,
  Div4WheelsOff varchar(10) DEFAULT NULL,
  Div4TailNum varchar(10) DEFAULT NULL,
  Div5Airport varchar(10) DEFAULT NULL,
  Div5AirportID int(11) DEFAULT NULL,
  Div5AirportSeqID int(11) DEFAULT NULL,
  Div5WheelsOn varchar(10) DEFAULT NULL,
  Div5TotalGTime varchar(10) DEFAULT NULL,
  Div5LongestGTime varchar(10) DEFAULT NULL,
  Div5WheelsOff varchar(10) DEFAULT NULL,
  Div5TailNum varchar(10) DEFAULT NULL
) DEFAULT CHARSET=latin1"

killall -9 tikv-importer
killall -9 tidb-lightning

cd /mnt/evo970
./tidb/tidb-latest-linux-amd64/bin/tikv-importer --config tikv-importer.toml &
./tidb/tidb-latest-linux-amd64/bin/tidb-lightning -config tidb-lightning.toml
```

2. What did you expect to see?

Success!

3. What did you see instead?

```
morgo@ryzen:~/bin$ time lightning-import-ontime 
tikv-importer: no process found
tidb-lightning: no process found
Error: not a valid TiDB version: 5.7.25-TiDB-v3.0.0-beta-211-g09beefbe0-dirty

real	0m0.329s
user	0m0.040s
sys	0m0.037s
```

4. Versions of the cluster

    - TiDB-Lightning version (run `tidb-lightning -V`):

        ```
Release Version: v2.1.5-3-g9c6a5f6
Git Commit Hash: 9c6a5f60dc765840cc09b4bcaa1149e93b3fa261
Git Branch: master
UTC Build Time: 2019-03-11 03:34:29
Go Version: go version go1.11.2 linux/amd64
        ```
        
    - TiKV-Importer version (run `tikv-importer -V`)

        ```
TiKV Importer 3.0.0-beta
        ```
        
    - TiKV version (run `tikv-server -V`):

        ```
TiKV 3.0.0-beta
        ```

    - TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

        ```
mysql> select tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v3.0.0-beta-211-g09beefbe0-dirty
Git Commit Hash: 09beefbe045011e3c77608c9ed33da87c11efa94
Git Branch: master
UTC Build Time: 2019-03-13 05:38:59
GoVersion: go version go1.11.3 linux/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false
1 row in set (0.00 sec)
        ```

    - Other interesting information (system version, hardware config, etc):

        >
        >

5. Operation logs
   - Please upload `tidb-lightning.log` for TiDB-Lightning if possible
   - Please upload `tikv-importer.log` from TiKV-Importer if possible
   - Other interesting logs
   
6. Configuration of the cluster and the task
   - `tidb-lightning.toml` for TiDB-Lightning if possible
   - `tikv-importer.toml` for TiKV-Importer if possible
   - `inventory.ini` if deployed by Ansible

7. Screenshot/exported-PDF of Grafana dashboard or metrics' graph in Prometheus for TiDB-Lightning if possible

### PR DESCRIPTION
**Pull request by lonng**
Signed-off-by: Lonng <chris@lonng.org>

### What problem does this PR solve? 

Fix #140 lightning breaks on tidb-master

### What is changed and how it works?

Add a new case for handling version format like `5.7.25-TiDB-v3.0.0-beta-211-g09beefbe0-dirty`

### Check List 

Tests 

 - Unit test

Side effects

N/A

Related changes

N/A

### CODE REVIEW

**Comment by lonng** [comment]
@kennytm @GregoryIan PTAL

File: lightning/restore/restore.go [hunk 1]
```diff
@@ -908,13 +908,17 @@ func extractTiDBVersion(version string) (*semver.Version, error) {
...
+	//                               ^~~~^
 	// The version is generated by `git describe --tags` on the TiDB repository.
 	versions := strings.Split(version, "-")
 	end := len(versions)
 	switch end {
 	case 3, 4:
 	case 5, 6:
 		end -= 2
+	case 7:
+		end -= 3
```
Comment (GregoryIan) [inline lightning/restore/restore.go]: is the version format fixed?

Comment (GregoryIan) [reply lightning/restore/restore.go]: I think what can be determined is `5.x.x-TiDB-{tidb version}-xxxxxxx`, it's better to extract front three item

Comment (lonng) [reply lightning/restore/restore.go]: The version format was generated by `git describe --tags --dirty --always`, so I think it will be a fixed format.

**Comment by kennytm** [comment]
I suggest we instead check if the last word is `dirty`, and slice it off it true. It is possible to create the version `5.7.25-TiDB-v3.0.0-dirty` for instance.

**Comment by lonng** [comment]
> I suggest we instead check if the last word is `dirty`, and slice it off it true. It is possible to create the version `5.7.25-TiDB-v3.0.0-dirty` for instance.

@kennytm Like this `versions := strings.Split(strings.TrimSuffix(version, "-dirty"), "-")`?

File: lightning/restore/restore.go [hunk 2]
```diff
@@ -908,13 +908,17 @@ func extractTiDBVersion(version string) (*semver.Version, error) {
 	//                               ^~~~^
 	// version format: "5.7.10-TiDB-v2.0.7"
 	//                               ^~~~^
+	// version format: "5.7.25-TiDB-v3.0.0-beta-211-g09beefbe0-dirty"
+	//                               ^~~~^
```
Comment (amyangfei) [inline lightning/restore/restore.go]:
```
	// version format: "5.7.25-TiDB-v3.0.0-beta-211-g09beefbe0-dirty"
	//                               ^~~~~~~~~^
```

Comment (lonng) [reply lightning/restore/restore.go]: addressed

**Comment by GregoryIan** [comment]
LGTM

**Pull request merged**